src/core/profiling/endoscope_backend.c

```diff
@@ -0,0 +1,577 @@
+          if (mytask->log_head == GRPC_ENDO_EMPTY) {
+            mytask->log_head = atom_index;
+          } else {
+            base->atom_pool[mytask->log_tail].next_atom = atom_index;
+          }
+          mytask->log_tail = atom_index;
+        }
+      }
+    }
+  }
+}
+
+static void grpc_endo_midpoint(grpc_endo_base *base, const char *name, const char *file, const gpr_int32 line,
+    const char* function_name, gpr_int64 cycle_event, GRPC_ENDO_INDEX marker_type, GRPC_ENDO_INDEX atom_type) {
+  GRPC_ENDO_INDEX marker_id;
+  grpc_endo_marker *mymarker;
+  gpr_uint64 thread_id;
+  GRPC_ENDO_INDEX thread_index;
+  grpc_endo_thread *mythread;
+
+  marker_id = grpc_endo_get_or_create_marker(base, line, name);
+  if (marker_id == GRPC_ENDO_EMPTY) {
+    grpc_endo_warning_print("grpc_endo_midpoint: cannot create marker item");
+    return;  /* only when out of markers */
+  }
+
+  mymarker = &(base->marker_pool[marker_id]);
+  if (mymarker->name == NULL) {  /* created new marker if not found */
+#ifdef GRPC_ENDO_COPY_MARKER_NAME
+      mymarker->name = strdup(name);
+#else
+      mymarker->name = name;
+#endif
+    mymarker->type = marker_type;
+    mymarker->file = file;
+    mymarker->line = line;
+    mymarker->function_name = function_name;
+    mymarker->cycle_created = cycle_event;
+  }
+
+  thread_id = gpr_thd_currentid();
+  thread_index = grpc_endo_get_thread(base, thread_id);
+  if (thread_index == GRPC_ENDO_EMPTY) {
+    grpc_endo_warning_print2("grpc_endo_midpoint cannot find thread item",
+             "midpoint not in scope", mymarker->name);
+    return;  /* only when thread invalid */
+  }
+  mythread = &(base->thread_pool[thread_index]);
+
+  if (mythread->task_active == GRPC_ENDO_EMPTY) {  /* top level - invalid */
+    grpc_endo_warning_print2("grpc_endo_midpoint no active task on current thread",
+             "midpoint not in scope", mymarker->name);
+    return;
+  } else {  /* should be in task or scope */
+    grpc_endo_task *mytask = &(base->task_pool[mythread->task_active]);
+    GRPC_ENDO_INDEX atom_index = grpc_endo_create_atom(base);
+    if (atom_index == GRPC_ENDO_EMPTY) {
+      grpc_endo_warning_print("grpc_endo_midpoint cannot create atom item");
+      return;  /* only when out of atoms */
+    } else {
+      grpc_endo_atom *myatom = &(base->atom_pool[atom_index]);
+      myatom->cycle = cycle_event;
+      myatom->type = atom_type;
+      myatom->param = marker_id;
+      if (mytask->log_head == GRPC_ENDO_EMPTY) {
+        mytask->log_head = atom_index;
+      } else {
+        base->atom_pool[mytask->log_tail].next_atom = atom_index;
+      }
+      mytask->log_tail = atom_index;
+    }
+  }
+}
+
+void grpc_endo_event(grpc_endo_base *base, const char *name, const char *file, const gpr_int32 line,
+                   const char* function_name, gpr_int64 cycle_event) {
+  grpc_endo_midpoint(base, name, file, line, function_name, cycle_event,
+                      3 /* EndoMarkerPB::EVENT */ , 
```

Comment:
maybe split these hard coded values into something like endo_proto.h:

`#define GRPC_ENDO_MARKER_PB_EVENT 3`

with a pointer to the .proto that defines them in a comment